### PR TITLE
[flang] IEEE_ARITHMETIC and IEEE_EXCEPTIONS intrinsic module procedures

### PR DESCRIPTION
**Pull request by vdonaldson**
Implement a selection of intrinsic module procedures that involve exceptions.

 - IEEE_GET_FLAG
 - IEEE_GET_HALTING_MODE
 - IEEE_GET_MODES
 - IEEE_GET_STATUS
 - IEEE_LOGB
 - [f23] IEEE_MAX, IEEE_MAX_MAG, IEEE_MAX_NUM, IEEE_MAX_NUM_MAG
 - [f23] IEEE_MIN, IEEE_MIN_MAG, IEEE_MIN_NUM, IEEE_MIN_NUM_MAG
 - IEEE_QUIET_EQ, IEEE_QUIET_GE, IEEE_QUIET_GT,
 - IEEE_QUIET_LE, IEEE_QUIET_LT, IEEE_QUIET_NE
 - IEEE_SET_FLAG
 - IEEE_SET_HALTING_MODE
 - IEEE_SET_MODES
 - IEEE_SET_STATUS
 - IEEE_SIGNALING_EQ, IEEE_SIGNALING_GE, IEEE_SIGNALING_GT,
 - IEEE_SIGNALING_LE, IEEE_SIGNALING_LT, IEEE_SIGNALING_NE
 - IEEE_SUPPORT_FLAG
 - IEEE_SUPPORT_HALTING

### CODE REVIEW

File: flang/runtime/exceptions.cpp [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+#include <cfenv>
+
+namespace Fortran::runtime {
+
+extern "C" {
+
+std::int32_t RTNAME(MapException)(int32_t except) {
+  Terminator terminator{__FILE__, __LINE__};
+
+  static constexpr int32_t mask = _FORTRAN_RUNTIME_IEEE_INVALID |
```
Comment (klausler) [inline flang/runtime/exceptions.cpp:23]: Please use braced initialization in the runtime.

File: flang/runtime/exceptions.cpp [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+      _FORTRAN_RUNTIME_IEEE_DENORM | _FORTRAN_RUNTIME_IEEE_DIVIDE_BY_ZERO |
+      _FORTRAN_RUNTIME_IEEE_OVERFLOW | _FORTRAN_RUNTIME_IEEE_UNDERFLOW |
+      _FORTRAN_RUNTIME_IEEE_INEXACT;
+  if (except != (except & mask))
```
Comment (klausler) [inline flang/runtime/exceptions.cpp:27]: Please always use braces on `if`, `while`, `for`, &c. in the runtime.

**Review by klausler — APPROVED** [review]

**Review by psteinfeld — APPROVED** [review]
All builds and tests correctly for me.  But I notice that one of the check builds is failing.  Please fix that before merging.

File: flang/runtime/exceptions.cpp (hunk 2 above)
Comment (vdonaldson) [reply flang/runtime/exceptions.cpp]: Done

**Comment by vdonaldson** [comment]
> All builds and tests correctly for me. But I notice that one of the check builds is failing. Please fix that before merging.

The test failures could be explained as nondeterminism in the mlir instruction order, which in turn might be dependent on which passes are run on the IR. Let's see if converting some `CHECK` directives to `CHECK-DAG` directives helps.

**Review by jeanPerier — APPROVED** [review]
Looks great

**Pull request merged**